rafthttp/stream.go

```diff
@@ -0,0 +1,388 @@
+package rafthttp
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"path"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/coreos/etcd/etcdserver/stats"
+	"github.com/coreos/etcd/pkg/types"
+	"github.com/coreos/etcd/raft/raftpb"
+)
+
+type streamType string
+
+const (
+	streamTypeGeneral streamType = "general"
+	streamTypeMsgApp  streamType = "msgapp"
+
+	streamBufSize = 4096
+)
+
+var (
+	heartbeatMessage = raftpb.Message{Type: raftpb.MsgHeartbeat}
+)
+
+func isHeartbeatMessage(m raftpb.Message) bool {
+	return m.Type == raftpb.MsgHeartbeat && m.From == 0 && m.To == 0
+}
+
+type encoder interface {
+	// encode encodes the given message to an output stream.
+	encode(m raftpb.Message) error
+}
+
+type decoder interface {
+	// decode decodes the message from an input stream.
+	decode() (raftpb.Message, error)
+}
+
+type outgoingStream struct {
+	encoder
+	http.Flusher
+	io.Closer
+}
+
+type streamWriter struct {
+	mu      sync.Mutex // guard field working
+	working bool
+
+	msgc    chan raftpb.Message
+	streamc chan *outgoingStream
+	stopc   chan struct{}
+	done    chan struct{}
+}
+
+func startStreamWriter() *streamWriter {
+	w := &streamWriter{
+		msgc:    make(chan raftpb.Message, streamBufSize),
+		streamc: make(chan *outgoingStream),
+		stopc:   make(chan struct{}),
+		done:    make(chan struct{}),
+	}
+	go w.run()
+	return w
+}
+
+func (cw *streamWriter) run() {
+	var msgc chan raftpb.Message
+	var heartbeatc <-chan time.Time
+	var stream *outgoingStream
+	tickc := time.Tick(ConnReadTimeout / 3)
+	reset := func(nstream *outgoingStream) {
+		if stream != nil {
+			stream.Close()
+		}
+		if nstream != nil {
+			heartbeatc, msgc = tickc, cw.msgc
+		} else {
+			heartbeatc, msgc = nil, nil
+		}
+		stream = nstream
+		cw.mu.Lock()
+		cw.working = nstream != nil
+		cw.mu.Unlock()
+	}
+	for {
+		select {
+		case <-heartbeatc:
+			if err := stream.encode(heartbeatMessage); err != nil {
+				log.Printf("rafthttp: encountered error heartbeating on stream: %v", err)
+				reset(nil)
+				continue
+			}
+			stream.Flush()
+		case m := <-msgc:
+			if err := stream.encode(m); err != nil {
+				log.Printf("rafthttp: encountered error encoding to stream: %v", err)
+				reset(nil)
+				continue
+			}
+			stream.Flush()
+		case stream := <-cw.streamc:
+			reset(stream)
+		case <-cw.stopc:
+			if stream != nil {
+				stream.Close()
+			}
+			close(cw.done)
+			return
+		}
+	}
+}
+
+func (cw *streamWriter) isWorking() bool {
```

Comment:
it is a writer/server. you give connection to it, and it will manage it for you then. Let me do the change we discussed first, which makes it clear.